tests/parse_metadata.py

```diff
@@ -0,0 +1,264 @@
+        data: dict[str, object] = tomli.load(f).get("tool", {}).get("stubtest", {})
+
+    skipped = data.get("skipped", False)
+    apt_dependencies = data.get("apt_dependencies", [])
+    brew_dependencies = data.get("brew_dependencies", [])
+    choco_dependencies = data.get("choco_dependencies", [])
+    extras = data.get("extras", [])
+    ignore_missing_stub = data.get("ignore_missing_stub", True)
+    specified_platforms = data.get("platforms", ["linux"])
+
+    assert type(skipped) is bool
+    assert type(ignore_missing_stub) is bool
+
+    # It doesn't work for type-narrowing if we use a for loop here...
+    assert _is_list_of_strings(specified_platforms)
+    assert _is_list_of_strings(apt_dependencies)
+    assert _is_list_of_strings(brew_dependencies)
+    assert _is_list_of_strings(choco_dependencies)
+    assert _is_list_of_strings(extras)
+
+    unrecognised_platforms = set(specified_platforms) - _STUBTEST_PLATFORM_MAPPING.keys()
+    assert not unrecognised_platforms, f"Unrecognised platforms specified for {distribution!r}: {unrecognised_platforms}"
+
+    for platform, dep_key in _STUBTEST_PLATFORM_MAPPING.items():
+        if platform not in specified_platforms:
+            assert dep_key not in data, (
+                f"Stubtest is not run on {platform} in CI for {distribution!r}, "
+                f"but {dep_key!r} are specified in METADATA.toml"
+            )
+
+    return StubtestSettings(
+        skipped=skipped,
+        apt_dependencies=apt_dependencies,
+        brew_dependencies=brew_dependencies,
+        choco_dependencies=choco_dependencies,
+        extras=extras,
+        ignore_missing_stub=ignore_missing_stub,
+        platforms=specified_platforms,
+    )
+
+
+@final
+@dataclass(frozen=True)
+class StubMetadata:
+    """The metadata for a single stubs distribution.
+
+    Don't construct instances directly; use the `read_metadata` function.
+    """
+
+    version: str
+    requires: Annotated[list[str], "The raw requirements as listed in METADATA.toml"]
+    extra_description: str | None
+    stub_distribution: Annotated[str, "The name under which the distribution is uploaded to PyPI"]
+    obsolete_since: Annotated[str, "A string representing a specific version"] | None
+    no_longer_updated: bool
+    uploaded_to_pypi: Annotated[bool, "Whether or not a distribution is uploaded to PyPI"]
+    stubtest_settings: StubtestSettings
+
+
+_KNOWN_METADATA_FIELDS: Final = frozenset(
+    {"version", "requires", "extra_description", "stub_distribution", "obsolete_since", "no_longer_updated", "upload", "tool"}
+)
+_KNOWN_METADATA_TOOL_FIELDS: Final = {
+    "stubtest": {
+        "skip",
+        "apt_dependencies",
+        "brew_dependencies",
+        "choco_dependencies",
+        "extras",
+        "ignore_missing_stub",
+        "platforms",
+    }
+}
+_DIST_NAME_RE: Final = re.compile(r"^[a-z0-9]([a-z0-9._-]*[a-z0-9])?$", re.IGNORECASE)
+
+
+@cache
+def read_metadata(distribution: str) -> StubMetadata:
```

Comment:
Idk if you wanna add them here or in a future improvement PR. But last time I played around with metadata validation, I figured there was a bunch of validation missing. Here's all of 'em here if you want inspiration: https://github.com/Avasam/typeshed/pull/13/files#diff-dcd1108b72e2746cfbed1553d3af6dec2f0623b2c1014dadb646c8577d64e046R77

Reply:
Oh, I didn't realise you'd already put so much work into this -- you should have said!

Though I prefer some of the choices made in my PR ;) (`dataclasses` > `TypedDict`s, in my opinion)

Reply:
I agree with the choices you made. I was just loosely playing around with some ideas. Nothing too serious.

Reply:
Still, y'know, feel free to say "Hey, I've already been poking around at this problem!" in cases like this :)

In terms of METADATA.toml validation, I can't see _much_ in your PR that isn't covered in this PR -- but if I'm missing something, feel free to add it in a followup PR!